library/std/src/sys/fs/windows.rs

```diff
@@ -1233,74 +1233,147 @@ pub fn unlink(path: &WCStr) -> io::Result<()> {
+            let mut file_attribute_tag_info: MaybeUninit<c::FILE_ATTRIBUTE_TAG_INFO> =
+                MaybeUninit::uninit();
+
+            let result = unsafe {
+                cvt(c::GetFileInformationByHandleEx(
+                    handle.as_raw_handle(),
+                    c::FileAttributeTagInfo,
+                    file_attribute_tag_info.as_mut_ptr().cast(),
+                    mem::size_of::<c::FILE_ATTRIBUTE_TAG_INFO>().try_into().unwrap(),
+                ))
             };
-            let offset: u32 = offset_of!(c::FILE_RENAME_INFO, FileName).try_into().unwrap();
-            let struct_size = offset + new_len_without_nul_in_bytes + 2;
-            let layout =
-                Layout::from_size_align(struct_size as usize, align_of::<c::FILE_RENAME_INFO>())
-                    .unwrap();
-
-            // SAFETY: We allocate enough memory for a full FILE_RENAME_INFO struct and a filename.
-            let file_rename_info;
-            unsafe {
-                file_rename_info = alloc(layout).cast::<c::FILE_RENAME_INFO>();
-                if file_rename_info.is_null() {
-                    return Err(io::ErrorKind::OutOfMemory.into());
+
+            if let Err(err) = result {
+                if err.raw_os_error() == Some(c::ERROR_INVALID_PARAMETER as _)
+                    || err.raw_os_error() == Some(c::ERROR_INVALID_FUNCTION as _)
+                {
+                    // `GetFileInformationByHandleEx` documents that not all underlying drivers support all file information classes.
+                    // Since we know we passed the correct arguments, this means the underlying driver didn't understand our request;
+                    // `MoveFileEx` proceeds by reopening the file without inhibiting reparse point behavior.
+                    None
+                } else {
+                    Some(Err(err))
+                }
+            } else {
+                // SAFETY: The struct has been initialized by GetFileInformationByHandleEx
+                let file_attribute_tag_info = unsafe { file_attribute_tag_info.assume_init() };
+                let file_type = FileType::new(
+                    file_attribute_tag_info.FileAttributes,
+                    file_attribute_tag_info.ReparseTag,
+                );
+
+                if file_type.is_symlink() {
+                    // The file is a mount point, junction point or symlink so
+                    // don't reopen the file so that the link gets renamed.
+                    Some(Ok(handle))
+                } else {
+                    // Otherwise reopen the file without inhibiting reparse point behavior.
+                    None
                 }
+            }
+        }
+        // The underlying driver may not support `FILE_FLAG_OPEN_REPARSE_POINT`: Retry without it.
+        Err(err) if err.raw_os_error() == Some(c::ERROR_INVALID_PARAMETER as _) => None,
+        Err(err) => Some(Err(err)),
+    }
+    .unwrap_or_else(|| create_file(0, 0))?;
```

Comment:
Looking at the error codes that are handled I'm worry this doesn't account for reparse points that cannot be opened, not because they're aren't support but because they cannot be interpreted by `CreateFile`. E.g. app execution aliases aren't actually links in the filesystem sense and can only be interpreted by `CreateProcess`. We should still be able to rename them though.